### ISSUE TITLE
Pyright is incorrectly handling overload functions

### ISSUE THREAD

**Issue by Pauan** [microsoft/pyright#11470]
**Describe the bug**

According to the [official Python spec for overload](https://typing.python.org/en/latest/spec/overload.html), the `@overload` function can be followed by a single non-overload function:

> In regular modules, a series of `@overload`-decorated definitions must be followed by exactly one non-`@overload`-decorated definition (for the same function/method).
>
> The `@overload`-decorated definitions are for the benefit of the type checker only, since they will be overwritten by the non-`@overload`-decorated definition, while the latter is used at runtime but ***should be ignored by a type checker***.

The spec also gives an example of an overloaded function, which does not specify types for the non-overloaded function:

```python
@overload
def utf8(value: None) -> None:
    pass
@overload
def utf8(value: bytes) -> bytes:
    pass
@overload
def utf8(value: unicode) -> bytes:
    pass
def utf8(value):
    <actual implementation>
```

The type checker should assume that the `@overload` functions accurately describe the entire type of the function, and the type checker should ignore the type of the non-overloaded function.

But pyright is incorrectly type checking the non-overload function:

```python
from typing import overload

class Foo:
    @overload
    def foo(self, value: int) -> int: ...

    @overload
    def foo(self, value: str) -> str: ...

    def foo(self, value):
        return value
```

When type checking that code in strict mode, pyright incorrectly gives the following errors:

```
__init__.py:10:19 - error: Type of parameter "value" is unknown (reportUnknownParameterType)

__init__.py:10:19 - error: Type annotation is missing for parameter "value" (reportMissingParameterType)
```

In order to suppress those errors, I need to add an `int | str` union type to the final non-overload function.

In that simple example it might seem okay, but with complicated types it can be very tricky indeed. Here is a real example from my project:

```python
@overload
def _make_item[A: JSON](self, id: str, default: A | None, item_class: type[Item[JSON]]) -> Item[A]: ...

@overload
def _make_item(self, id: str, default: dict[str, JSON] | None, item_class: type["ItemDict"]) -> "ItemDict": ...

@overload
def _make_item(self, id: str, default: list[dict[str, JSON]] | None, item_class: type["ItemList"]) -> "ItemList": ...

def _make_item(self, id, default, item_class):
    # Actual function body...
```

Trying to union the 3 different overloads is very annoying, and a union isn't even correct because of the generic.

In order to comply with the spec, pyright must ignore the type of the final non-overload function in the `@overload` chain.

I would even go further and suggest that pyright should error if the user tries to specify a type for the non-overload function, forcing the non-overload function to remain untyped.

This prevents the footgun where a user adds a type to the non-overload function and is then confused why it's not working.

Of course the *body* of the function should still be type-checked as usual, it is only the function arguments and return type that must remain untyped.

**VS Code extension or command-line**

I'm running pyright from the command line, [installed from pypi](https://pypi.org/project/pyright/). It is version 1.1.409.

**Comment by rchiodo** [comment]
I'm not sure I agree that the spec says the overload implementation shouldn't have type arguments:
https://typing.python.org/en/latest/spec/overload.html#implementation-consistency

I feel like that section contradicts the `ignore` comment further up.

**Comment by Pauan** [comment]
> @rchiodo I feel like that section contradicts the ignore comment further up.

I don't see any contradiction.

> If an overload implementation is defined, type checkers should validate that it is consistent with all of its associated overload signatures.
>
> In typing terms, this means the input signature of the implementation should be [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to the input signatures of all overloads, and the return type of all overloads should be assignable to the return type of the implementation.

It's saying that the implementation function must be type consistent with its overloads, but it's not saying that users must specify a union type for the implementation function.

If the user doesn't specify a type signature for the implementation function, then it is implied to be `Any`, and thus it is always assignable to the overloads, and is thus valid and should be accepted.

But just because it's implied to be `Any` doesn't mean you can call the function with any type! Because it says this...

> Only the overloads (the @overload-decorated signatures) should be considered for matching purposes. The 
implementation, if provided, should be ignored for purposes of overload matching.

So even if the implementation function has a type of `Any`, that type is ignored and it only considers the type of the overloads.

This is consistent with the spec saying that "the implementation function is used at runtime but should be ignored by a type checker".

> The implementation should accept all potential sets of arguments that are accepted by the overloads and should produce all potential return types produced by the overloads.

In other words, it's not type checking the implementation function against its own type signature, it's always type checking against the type signature of the overloads. Its own signature is irrelevant.

----

So the only behavior that consistently aligns with the entirety of the spec is:

1. An implementation function does not contribute to the type of the function.

   The type of the function is only based on the overloads and nothing else.

   Even if the implementation specifies a type signature, it still doesn't contribute to the type of the function.

   Even if the user manually specifies `Any` for the implementation function (such as `def foo(x: Any, y: Any) -> Any`), that type is ignored, and the function type will still be based only on the overloads.

   This is clear from the definition of assignability, which has special rules [only for `@overload`](https://typing.python.org/en/latest/spec/callables.html#overloads).

   The assignability of an overloaded function only depends on the `@overload` types, not the implementation function type.

2. The implementation body is checked to be consistent with all of the overloads.

3. Implementations are allowed to not have a type signature. This is valid because they are implied to be `Any`. This is clear from the `utf8` example.

   But because it is ignored for type purposes it does *not* cause the function to have `Any` type. The function type is still based only on the overloads.

4. If the implementation *does* specify a type signature, the type signature is checked that it is consistent with all of the overloads.

**Comment by rchiodo** [comment]
This is the part that Pyright isn't enforcing:

> 3. Implementations are allowed to not have a type signature. This is valid because they are implied to be Any. This is clear from the utf8 example.

> But because it is ignored for type purposes it does not cause the function to have Any type. The function type is still based only on the overloads._

Pyright doesn't care that the function you want to check is an overload. In strict mode it checks _everything_.
